### ISSUE TITLE
Is optimizer.step() thread safe?

### ISSUE THREAD

**Issue by nilsjohanbjorck** [pytorch/pytorch#8149]
In the example of asynchronous training (examples/mnist_hogwild/train.py) a model is made shared via model.share_memory(). After that, the different threads simply call optimizer.step() asynchronously. In the documentation (https://pytorch.org/docs/master/notes/multiprocessing.html) it is recommended to use queues for passing tensors, but that is evidently not done in the example.

Is the optimizer.step() function guaranteed to be thread safe?

**Comment by soumith** [comment]
- "hogwild" throws out thread-safety out of the window -- so it's actually irrelevant for the example
- we start different processes (not threads)
- we dont have global semaphores for the optimizer to make it "safe" when working on the same shared parameters
- [tearing](https://stackoverflow.com/a/1607478) doesn't occur at the granularity of a float32 (on CPU or supported nvidia GPUs), so really bad things dont happen (unless maybe you train on dtype.float64 and on CUDA)

**Comment by Steamgjk** [comment]
I think I still have some concerns about the problem. Though tearing will not happen, but whether it will   cause read-write conflicts?

optimizer.step() needs to use gradients to update the model parameters, but two processes generate one replica of grads each, and they will simultaneously aggregate their gradients to the model parameters. Suppose the original value of the model parameter is 1.0, Process-0 generates a gradient 0.1, and Process-1 generates a gradient 0.2. They read the model parameter at the same time, then they both regard 1.0 as the original value and calculates a new value for model parameters. Say, Process-0 calculates 1-0.1=0.9, Process-1 calculates 1-0.2=0.8. Then they write back their values. If 0.8 comes later, the updated value will be 0.8 (1.0->0.9->0.8), otherwise, the updated value will be 0.9 (1.0->0.8->0.9). But actually, the effort of one worker process is lost. Then, no matter how many processes we launched simultaneously, the effort of most of them will be lost, right?

**Comment by Yanzzzzz** [comment]
> I think I still have some concerns about the problem. Though tearing will not happen, but whether it will cause read-write conflicts?
> 
> optimizer.step() needs to use gradients to update the model parameters, but two processes generate one replica of grads each, and they will simultaneously aggregate their gradients to the model parameters. Suppose the original value of the model parameter is 1.0, Process-0 generates a gradient 0.1, and Process-1 generates a gradient 0.2. They read the model parameter at the same time, then they both regard 1.0 as the original value and calculates a new value for model parameters. Say, Process-0 calculates 1-0.1=0.9, Process-1 calculates 1-0.2=0.8. Then they write back their values. If 0.8 comes later, the updated value will be 0.8 (1.0->0.9->0.8), otherwise, the updated value will be 0.9 (1.0->0.8->0.9). But actually, the effort of one worker process is lost. Then, no matter how many processes we launched simultaneously, the effort of most of them will be lost, right?
=====================================================================

I have the same confusion, and I think you are correct. I believe most of the threads updates will be ignored. 

Is anyone familiar with this can answer this?

**Comment by thomashirtz** [comment]
I am not familiar with this, but I am really interested if someone possess more information
@soumith would it be different if we were doing threading ? I am not sure I understand why 
Does someone have arguments for the message of @Steamgjk ?

I saw some people implementing the Hogwild! for the A3C using [this](https://github.com/MorvanZhou/pytorch-A3C/blob/master/shared_adam.py) kind of implementation. Your comment means that it is not really necessary soumith ?
